src/main/java/com/minecolonies/coremod/entity/ai/citizen/herders/AbstractEntityAIHerder.java

```diff
@@ -0,0 +1,534 @@
+
+        final boolean hasBreedingItem = worker.hasItemInInventory(getBreedingItem(), 0);
+
+        if (!searchForItemsInArea().isEmpty())
+        {
+            return HERDER_PICKUP;
+        }
+        else if (maxAnimals())
+        {
+            return HERDER_BUTCHER;
+        }
+        else if (numOfBreedableAnimals >= NUM_OF_ANIMALS_TO_BREED && hasBreedingItem)
+        {
+            return HERDER_BREED;
+        }
+        return START_WORKING;
+    }
+
+    /**
+     * Redirects the herder to their building.
+     *
+     * @return The next {@link AIState}.
+     */
+    private AIState startWorkingAtOwnBuilding()
+    {
+        worker.setLatestStatus(new TextComponentTranslation("com.minecolonies.coremod.status.herder.goingToHut"));
+        if (walkToBuilding())
+        {
+            return getState();
+        }
+        return PREPARING;
+    }
+
+    /**
+     * Prepares the herder for herding
+     *
+     * @return The next {@link AIState}.
+     */
+    private AIState prepareForHerding()
+    {
+        toolsNeeded.add(ToolType.AXE);
+        itemsNeeded.add(new ItemStack(getBreedingItem(), 2));
+
+        for (final ToolType tool : toolsNeeded)
+        {
+            if (checkForToolOrWeapon(tool))
+            {
+                return getState();
+            }
+        }
+
+        for (final ItemStack item : itemsNeeded)
+        {
+            checkOrRequestItemsAsynch(false, item);
+        }
+
+        return HERDER_DECIDE;
+    }
+
+    /**
+     * Butcher some animals (Preferably Adults) that the herder looks after.
+     *
+     * @return The next {@link AIState}.
+     */
+    private AIState butcherAnimals()
+    {
+        setDelay(DELAY_TWENTY);
+
+        if (!maxAnimals())
+        {
+            return HERDER_DECIDE;
+        }
+
+        if (!equipTool(ToolType.AXE))
+        {
+            return START_WORKING;
+        }
+
+        EntityAnimal animal = searchForAnimals().stream().filter(animalToButcher -> !animalToButcher.isChild()).findFirst().orElse(null);
+
+        if (animal == null)
+        {
+            animal = searchForAnimals().stream().findFirst().orElse(null);
+        }
+
+        butcherAnimal(animal);
+
+        if (animal != null && !animal.isEntityAlive())
+        {
+            incrementActionsDone();
+        }
+
+        return HERDER_BUTCHER;
+    }
+
+    /**
+     * Breed some animals together.
+     *
+     * @return The next {@link AIState}.
+     */
+    private AIState breedAnimals()
+    {
+        setDelay(DELAY_FOURTY);
+
+        final List<T> animals = new ArrayList<>(searchForAnimals());
+
+        final EntityAnimal animalOne = animals.stream().filter(animal -> animal.getGrowingAge() == 0).findAny().orElse(null);
+
+        if (animalOne == null)
+        {
+            return HERDER_DECIDE;
+        }
+
+        final EntityAnimal animalTwo = animals.stream().filter(animal ->
+          {
```

Comment:
this can be a helper function :)

Reply:
Sorry, i'm not sure what you mean by that one